Design/Instrument-2.0/Instrument_Ray_Tracer_2.md

```diff
@@ -0,0 +1,126 @@
+
+The intention is to stop developers needing access to the `Instrument` API and instead steer them towards using the new `Geometry` and `Beamline` layers as mentioned above. The layers provided by `Instrument 2.0` are much more efficient and the read performance is much better. It makes sense to start re-implementing code to make full use of the new layers and this move can help to optimise much of the code base. Also the `ComponentInfo` layer has been designed in such a way that the majority (if not all) of the functionality provided by `Instrument` is readily available via `ComponentInfo`.
+
+The goal of this document is to outline a plan for `InstrumentRayTracer 2.0` and to figure out how to move away from making any calls to the legacy API.
+ 
+## `InstrumentRayTracer 2.0` Methods
+
+#### Reuse of Original Code
+One idea for the implementation of `InstrumentRayTracer 2.0` is to have many of the old methods from `InstrumentRayTracer` carried over. This is would be a good way to proceed as it would hopefully mean that there will not be as many breaking changes to worry about. The methods to be replaced would be where the legacy API calls are made - i.e. anything that uses `Instrument` could be modified so that the new `Geometry` and `Beamline` layers are used instead. Methods headers that could be kept the same include:
+ * `void trace(const Kernel::V3D &dir) const`
+ * `void traceFromSample(const Kernel::V3D &dir) const`
+ * `Links getResults() const`
+ * `IDetector_const_sptr getDetectorResult() const`
+ * `void fireRay(Track &testRay) const`
+
+#### Possible New Methods
+Although the headers do not explicity have anything to do with `Instrument`, many of the method bodies make use of the member variable `m_instrument` to carry out various tasks. This variable is created and initialised in the constructor. From looking at the implementation in [`InstrumentRayTracer.cpp`](https://github.com/BhuvanBezawada/mantid/blob/master/Framework/Geometry/src/Objects/InstrumentRayTracer.cpp), it seems that the constructor is the one place where there is a dependence on `Instrument` and this is probably the one area that needs to be fully rewritten. 
+
+## The Constructor
+To rewrite the constructor it may be enough to pass in a `const` reference to a `ComponentInfo` object. This would ensure that the `InstrumentRayTracer` works as intended but there is a chance that undefined behaviour could arise if scoping is used.
+
+
+```c++
+// This example would work as intended
+InstrumentRayTracer2* ptr;
+auto ws = CreateWorkspace();
+ptr = new InstrumentRayTracer2(ws.componentInfo);
+ptr.trace(V3D(0,0,0));
+
+// With the following scoping, there would be undefined behaviour
+InstrumentRayTracer2* ptr;
+{
+  auto ws = CreateWorkspace();
+  ptr = new InstrumentRayTracer2(ws.componentInfo);
+}
+
+ptr.trace(V3D(0,0,0));
```

Comment:
This is certain undefined territory. const refs only extend the lifetime of local temporary objects. Does not apply to members.

Reply:
I agree. I prefer the latter two implementation suggestions. 

I think both suggestions have their strengths. The `ComponentInfo` implementation is nice in the case that you wish to trace multiple times for the same instrument. That implementation removes the need for the `ComponentInfo` object to be in scope for every trace call. 

However the classless implementation makes it very explicit what set of components you're tracing on. This would also make the usage simpler: the most common usage is to trace to get a detector. Currently this is two function calls. With a classless implementation I guess this would drop to one. Something like this?

```cpp
auto detector = InstrumentRayTracer2::traceDetector(compInfo, direction);
```